### DISCUSSION TITLE
How can I use `tsserver` locally from `node_modules`?

### DISCUSSION THREAD

**Discussion by noghartt** [joaotavora/eglot#987] [Q&A]
I'm a beginner with Emacs and I just try to use `eglot` as my LSP  client. Following this [setup](https://vxlabs.com/2022/06/12/typescript-development-with-emacs-tree-sitter-and-lsp-in-2022) I can run `eglot`, but I have a doubt now:

Is there a way to put eglot to run with `tsserver` from `node_modules` of my project instead of `PATH`? Because exists a `tsserver` on `node_modules/.bin` and I would like to use it.

**Comment by joaotavora** [comment]
I'd try `C-u M-x eglot` and figure out the relative path (which I _think_ is relative to the project's path) to the server and entering manually at the prompt.

Then after you're happy and verify the server works, you modify the Elisp variable `eglot-server-programs`, either locally, or buffer-locally, or dir-locally to contain the very same string that you entered manually in the step before.

Thereafter a simple `M-x eglot` (no `C-u`) in one of your project's buffers will (hopefully) do the right thing. 

If this answer is too terse, let's hope someone else chimes in, I don't have time to expand on it.

**Comment by zhaozhemin** [comment]
You can use [add-node-modules-path](https://github.com/codesuki/add-node-modules-path). It will automatically add `node_modules/.bin/` to your `PATH`.

**Comment by klandergren** [comment]
@noghartt my understanding is that, confusingly, `tsserver` is a ["language server"](https://github.com/microsoft/TypeScript/wiki/Standalone-Server-(tsserver)) but it does not currently implement Language Server Protocol (LSP) support to work with `eglot`. See [tsserver should implement the Language Server Protocol  #39459](https://github.com/microsoft/TypeScript/issues/39459) for more information.

[typescript-language-server](https://github.com/typescript-language-server/typescript-language-server), which is not an official TypeScript project but does wrap `tsserver`, is available for use with LSP clients like `eglot`.

Also, following @zhaozhemin's advice to use [add-node-modules-path](https://github.com/codesuki/add-node-modules-path) should work so that `eglot` can find `typescript-language-server`.

Here is an extract of my configuration that assumes you:

 * have installed `typescript-language-server` in your project via your JavaScript package manager of choice (e.g. `npm install typescript-language-server`)
 * are running emacs 29
 * have the corresponding Emacs packages installed (if not, try `M-x package-install`)
 * have `add-node-modules-path` in a directory `lisp/` relative from your Emacs configuration directory

```elisp
(use-package eglot
  :defer t)
  
(use-package add-node-modules-path
  :defer t

  :commands
  (add-node-modules-path)

  :load-path "lisp/")
  
(use-package typescript-ts-mode
  :defer t

  :mode
  ("\\.ts\\'" . typescript-ts-mode)     ; from `typescript-ts-mode.el`
  ("\\.tsx\\'" . tsx-ts-mode)           ; from `typescript-ts-mode.el`

  :config
  (progn
    (defun kl/ts-ts-mode-hook ()
      (add-node-modules-path)
      (eglot-ensure))

    (add-hook 'typescript-ts-mode-hook #'kl/ts-ts-mode-hook)
    (add-hook 'tsx-ts-mode-hook #'kl/ts-ts-mode-hook)))
```

You may need to tweak things to get it working for your set up. Good luck!
